# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1531422013 14400
#      Thu Jul 12 15:00:13 2018 -0400
# Node ID 02766a14c461f706ae47ae639b78242dbe8914d9
# Parent  c21edcb90cbf81b914bc9693b40f6d0eb7d0c9f0
add info about version numbers in etc/HACKING.md file

* HACKING.md: Describe the meaning of the Octave version number and
the Octave shared library version numbers and how to update them.

* libgui/module.mk, libinterp/module.mk, liboctave/module.mk:
Refer to etc/HACKING.md file instead of libtool manual for explanation
of how to update shared library version numbers.

* configure.ac: Refer to etc/HACKING.md file for explanation of how to
update Octave version number.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -23,16 +23,20 @@ AC_PREREQ([2.65])
 AC_INIT([GNU Octave], [5.0.0], [https://octave.org/bugs.html], [octave],
         [https://www.gnu.org/software/octave/])
 
 ### Declare version numbers
 
 dnl Note that the version number is duplicated here and in AC_INIT because
 dnl AC_INIT requires it to be static, not computed from shell variables.
 
+## The description of the Octave version number in the etc/HACKING.md
+## explains how to update these numbers for release and development
+## versions.
+
 OCTAVE_MAJOR_VERSION=5
 OCTAVE_MINOR_VERSION=0
 OCTAVE_PATCH_VERSION=0
 
 dnl PACKAGE_VERSION is set by the AC_INIT VERSION argument.
 OCTAVE_VERSION="$PACKAGE_VERSION"
 
 OCTAVE_COPYRIGHT="Copyright (C) 2018 John W. Eaton and others."
diff --git a/etc/HACKING.md b/etc/HACKING.md
--- a/etc/HACKING.md
+++ b/etc/HACKING.md
@@ -295,16 +295,104 @@ An overview of the directory layout of O
 
 - `test`
      tests for the interpreter.
   - `*.tst`
        fixed tests for the interpreter.
   - `fntests.m`
        script to run function tests embedded in C++ and .m files.
 
+Release Numbering
+-----------------
+
+Since version 5, Octave uses the following rules for release numbering:
+
+  Version Dev Phase       When
+
+  5.0.0   (experimental)  active development of Octave 5 on default branch
+  5.0.1   (pre-release)   stabilization period of Octave 5 on stable branch
+  6.0.0   (experimental)  active development of Octave 6 on default branch
+  5.1.0   (release)       first release of Octave 5 from stable branch
+  5.1.1   (pre-release)   bug fixing on stable branch after 5.1.0 release
+  5.2.0   (release)       second release of Octave 5 from stable branch
+  5.2.1   (pre-release)   bug fixing on stable branch after 5.2.0 release
+  ...
+
+To summarize, the first release of Octave 5 will be Octave 5.1.0 while
+development snapshots will be Octave 5.0.0 and snapshots from the
+release branch Octave 5.n.1.
+
+With this numbering scheme:
+
+  * Any version X.0.0 means "this is an experimental development
+    version".
+
+  * Any version X.Y.1 means, "this is a pre-release version meant
+    for bug fixing and testing".
+
+  * Any version X.Y.0 with Y != 0 means "this is a released version".
+
+Shared Library Versioning
+-------------------------
+
+Version numbers for the liboctave, liboctinterp, and liboctgui shared
+libraries are set in the module.mk files in the top-level directory for
+each library using the variables
+
+  %canon_reldir%_%canon_reldir%_current
+  %canon_reldir%_%canon_reldir%_revision
+  %canon_reldir%_%canon_reldir%_age
+
+The rules for updating these version numbers are:
+
+  * Start with version information of ‘0:0:0’ for each libtool library.
+
+  * Update the version information only immediately before a public
+    release of your software.  More frequent updates are unnecessary,
+    and only guarantee that the current interface number gets larger
+    faster.
+
+  * If the library source code has changed at all since the last update,
+    then increment revision (‘c:r:a’ becomes ‘c:r+1:a’).
+
+  * If any interfaces have been added, removed, or changed since the
+    last update, increment current, and set revision to 0.
+
+  * If any interfaces have been added since the last public release,
+    then increment age.
+
+  * If any interfaces have been removed or changed since the last public
+    release, then set age to 0.
+
+Never try to set the interface numbers so that they correspond to the
+Octave version number.  This is an abuse that only fosters
+misunderstanding of the purpose of library versions.
+
+The following explanation may help to understand the above rules a bit
+better: consider that there are three possible kinds of reactions from
+users of your library to changes in a shared library:
+
+  * Programs using the previous version may use the new version as
+    drop-in replacement, and programs using the new version can also
+    work with the previous one.  In other words, no recompiling nor
+    relinking is needed.  In this case, bump revision only, don’t touch
+    current nor age.
+
+  * Programs using the previous version may use the new version as
+    drop-in replacement, but programs using the new version may use APIs
+    not present in the previous one.  In other words, a program linking
+    against the new version may fail with unresolved symbols if linking
+    against the old version at runtime: set revision to 0, bump current
+    and age.
+
+  * Programs may need to be changed, recompiled, and relinked in order
+    to use the new version.  Bump current, set revision and age to 0.
+
+These guidelines also appear in the automake manual.
+
 
 ################################################################################
 
 Copyright (C) 2009-2018 John W. Eaton
 
 This file is part of Octave.
 
 Octave is free software: you can redistribute it and/or modify it
diff --git a/libgui/module.mk b/libgui/module.mk
--- a/libgui/module.mk
+++ b/libgui/module.mk
@@ -57,17 +57,19 @@ nodist_%canon_reldir%_liboctgui_la_SOURC
 %canon_reldir%_liboctgui_la_LIBADD = \
   %reldir%/qterminal/libqterminal.la \
   %reldir%/src/libgui-src.la \
   %reldir%/graphics/libgui-graphics.la \
   libinterp/liboctinterp.la \
   liboctave/liboctave.la \
   $(LIBOCTGUI_LINK_DEPS)
 
-# Increment these as needed and according to the rules in the libtool manual:
+## Increment the following version numbers as needed and according
+## to the rules in the etc/HACKING.md file:
+
 %canon_reldir%_liboctgui_current = 3
 %canon_reldir%_liboctgui_revision = 0
 %canon_reldir%_liboctgui_age = 0
 
 %canon_reldir%_liboctgui_version_info = $(%canon_reldir%_liboctgui_current):$(%canon_reldir%_liboctgui_revision):$(%canon_reldir%_liboctgui_age)
 
 %canon_reldir%_liboctgui_la_LDFLAGS = \
   -version-info $(%canon_reldir%_liboctgui_version_info) \
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -157,17 +157,19 @@ nodist_%canon_reldir%_liboctinterp_la_SO
   liboctave/liboctave.la \
   $(LIBOCTINTERP_LINK_DEPS)
 
 if AMCOND_BUILD_EXTERNAL_LIBXERBLA
   %canon_reldir%_liboctinterp_la_LIBADD += \
     liboctave/external/blas-xtra/libxerbla.la
 endif
 
-# Increment these as needed and according to the rules in the libtool manual:
+## Increment the following version numbers as needed and according
+## to the rules in the etc/HACKING.md file:
+
 %canon_reldir%_liboctinterp_current = 5
 %canon_reldir%_liboctinterp_revision = 0
 %canon_reldir%_liboctinterp_age = 0
 
 %canon_reldir%_liboctinterp_version_info = $(%canon_reldir%_liboctinterp_current):$(%canon_reldir%_liboctinterp_revision):$(%canon_reldir%_liboctinterp_age)
 
 %canon_reldir%_liboctinterp_la_LDFLAGS = \
   -version-info $(%canon_reldir%_liboctinterp_version_info) \
diff --git a/liboctave/module.mk b/liboctave/module.mk
--- a/liboctave/module.mk
+++ b/liboctave/module.mk
@@ -81,17 +81,19 @@ nodist_%canon_reldir%_%canon_reldir%_la_
   %reldir%/liboctave-build-info.cc \
   %reldir%/version.cc \
   %reldir%/version.h
 
 %canon_reldir%_%canon_reldir%_la_LIBADD += \
   libgnu/libgnu.la \
   $(LIBOCTAVE_LINK_DEPS)
 
-# Increment these as needed and according to the rules in the libtool manual:
+## Increment the following version numbers as needed and according
+## to the rules in the etc/HACKING.md file:
+
 %canon_reldir%_%canon_reldir%_current = 5
 %canon_reldir%_%canon_reldir%_revision = 0
 %canon_reldir%_%canon_reldir%_age = 0
 
 %canon_reldir%_%canon_reldir%_version_info = $(%canon_reldir%_%canon_reldir%_current):$(%canon_reldir%_%canon_reldir%_revision):$(%canon_reldir%_%canon_reldir%_age)
 
 %canon_reldir%_%canon_reldir%_la_LDFLAGS = \
   -version-info $(%canon_reldir%_%canon_reldir%_version_info) \
